File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteInstanceRecord.java
package com.krystianwsul.checkme.firebase.records;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import com.krystianwsul.checkme.domainmodel.DomainFactory;
import com.krystianwsul.checkme.firebase.json.InstanceJson;
import com.krystianwsul.checkme.utils.CustomTimeKey;
import com.krystianwsul.checkme.utils.ScheduleKey;
import com.krystianwsul.checkme.utils.time.Date;
import com.krystianwsul.checkme.utils.time.HourMinute;
import com.krystianwsul.checkme.utils.time.TimePair;

import junit.framework.Assert;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class RemoteInstanceRecord extends RemoteRecord {
    private static final Pattern sHourMinutePattern = Pattern.compile("^(\\d\\d\\d\\d)-(\\d?\\d)-(\\d?\\d)-(\\d?\\d)-(\\d?\\d)$");
    private static final Pattern sCustomTimePattern = Pattern.compile("^(\\d\\d\\d\\d)-(\\d?\\d)-(\\d?\\d)-(.+)$");

    @NonNull
    private final DomainFactory mDomainFactory;

    @NonNull
    private final RemoteTaskRecord mRemoteTaskRecord;

    @NonNull
    private final InstanceJson mInstanceJson;

    @NonNull
    private final ScheduleKey mScheduleKey;

    RemoteInstanceRecord(boolean create, @NonNull DomainFactory domainFactory, @NonNull RemoteTaskRecord remoteTaskRecord, @NonNull InstanceJson instanceJson, @NonNull ScheduleKey scheduleKey) {
        super(create);

        mDomainFactory = domainFactory;
        mRemoteTaskRecord = remoteTaskRecord;
        mInstanceJson = instanceJson;
        mScheduleKey = scheduleKey;
    }

    @NonNull
    @Override
    protected String getKey() {
        return mRemoteTaskRecord.getKey() + "/instances/" + scheduleKeyToString(mDomainFactory, mRemoteTaskRecord.getProjectId(), getScheduleKey());
    }

    @NonNull
    ScheduleKey getScheduleKey() {
        return mScheduleKey;
    }

    @NonNull
    public static String scheduleKeyToString(@NonNull DomainFactory domainFactory, @NonNull String projectId, @NonNull ScheduleKey scheduleKey) {
        String key = scheduleKey.getScheduleDate().getYear() + "-" + scheduleKey.getScheduleDate().getMonth() + "-" + scheduleKey.getScheduleDate().getDay();
        if (scheduleKey.getScheduleTimePair().getCustomTimeKey() != null) {
            Assert.assertTrue(scheduleKey.getScheduleTimePair().getHourMinute() == null);

            key += "-" + domainFactory.getRemoteCustomTimeId(projectId, scheduleKey.getScheduleTimePair().getCustomTimeKey());
        } else {
            Assert.assertTrue(scheduleKey.getScheduleTimePair().getHourMinute() != null);

            key += "-" + scheduleKey.getScheduleTimePair().getHourMinute().getHour() + "-" + scheduleKey.getScheduleTimePair().getHourMinute().getMinute();
        }

        return key;
    }

    @NonNull
    static ScheduleKey stringToScheduleKey(@NonNull DomainFactory domainFactory, @NonNull String projectId, @NonNull String key) {
        Matcher hourMinuteMatcher = sHourMinutePattern.matcher(key);

        if (hourMinuteMatcher.matches()) {
            int year = Integer.valueOf(hourMinuteMatcher.group(1));
            int month = Integer.valueOf(hourMinuteMatcher.group(2));
            int day = Integer.valueOf(hourMinuteMatcher.group(3));
            int hour = Integer.valueOf(hourMinuteMatcher.group(4));
            int minute = Integer.valueOf(hourMinuteMatcher.group(5));

            return new ScheduleKey(new Date(year, month, day), new TimePair(new HourMinute(hour, minute)));
        } else {
            Matcher customTimeMatcher = sCustomTimePattern.matcher(key);
            Assert.assertTrue(customTimeMatcher.matches());

            int year = Integer.valueOf(customTimeMatcher.group(1));
            int month = Integer.valueOf(customTimeMatcher.group(2));
            int day = Integer.valueOf(customTimeMatcher.group(3));

            String customTimeId = customTimeMatcher.group(4);
            Assert.assertTrue(!TextUtils.isEmpty(customTimeId));

            CustomTimeKey customTimeKey = domainFactory.getCustomTimeKey(projectId, customTimeId);

            return new ScheduleKey(new Date(year, month, day), new TimePair(customTimeKey));
        }
    }

    @NonNull
    @Override
    protected InstanceJson getCreateObject() {
        return mInstanceJson;
    }

    @NonNull
    public String getTaskId() {
        return mRemoteTaskRecord.getId();
    }

    public Long getDone() {
        return mInstanceJson.getDone();
    }

    public int getScheduleYear() {
        return mScheduleKey.getScheduleDate().getYear();
    }

    public int getScheduleMonth() {
        return mScheduleKey.getScheduleDate().getMonth();
    }

    public int getScheduleDay() {
        return mScheduleKey.getScheduleDate().getDay();
    }

    @Nullable
    public String getScheduleCustomTimeId() {
        CustomTimeKey customTimeKey = mScheduleKey.getScheduleTimePair().getCustomTimeKey();
        if (customTimeKey != null) {
            return mDomainFactory.getRemoteCustomTimeId(mRemoteTaskRecord.getProjectId(), customTimeKey);
        } else {
            return null;
        }
    }

    @Nullable
    public Integer getScheduleHour() {
        HourMinute hourMinute = mScheduleKey.getScheduleTimePair().getHourMinute();
        if (hourMinute != null) {
            return hourMinute.getHour();
        } else {
            return null;
        }
    }

    @Nullable
    public Integer getScheduleMinute() {
        HourMinute hourMinute = mScheduleKey.getScheduleTimePair().getHourMinute();
        if (hourMinute != null) {
            return hourMinute.getMinute();
        } else {
            return null;
        }
    }

    @Nullable
    public Integer getInstanceYear() {
        return mInstanceJson.getInstanceYear();
    }

    @Nullable
    public Integer getInstanceMonth() {
        return mInstanceJson.getInstanceMonth();
    }

    @Nullable
    public Integer getInstanceDay() {
        return mInstanceJson.getInstanceDay();
    }

    @Nullable
    public String getInstanceCustomTimeId() {
        return mInstanceJson.getInstanceCustomTimeId();
    }

    @Nullable
    public Integer getInstanceHour() {
        return mInstanceJson.getInstanceHour();
    }

    @Nullable
    public Integer getInstanceMinute() {
        return mInstanceJson.getInstanceMinute();
    }

    public long getHierarchyTime() {
        return mInstanceJson.getHierarchyTime();
    }

    public void setInstanceYear(int instanceYear) {
        if (getInstanceYear() != null && getInstanceYear().equals(instanceYear))
            return;

        mInstanceJson.setInstanceYear(instanceYear);
        addValue(getKey() + "/instanceYear", instanceYear);
    }

    public void setInstanceMonth(int instanceMonth) {
        if (getInstanceMonth() != null && getInstanceMonth().equals(instanceMonth))
            return;

        mInstanceJson.setInstanceMonth(instanceMonth);
        addValue(getKey() + "/instanceMonth", instanceMonth);
    }

    public void setInstanceDay(int instanceDay) {
        if (getInstanceDay() != null && getInstanceDay().equals(instanceDay))
            return;

        mInstanceJson.setInstanceDay(instanceDay);
        addValue(getKey() + "/instanceDay", instanceDay);
    }

    public void setInstanceCustomTimeId(@Nullable String instanceCustomTimeId) {
        if (!TextUtils.isEmpty(getInstanceCustomTimeId()) && getInstanceCustomTimeId().equals(instanceCustomTimeId))
            return;

        mInstanceJson.setInstanceCustomTimeId(instanceCustomTimeId);
        addValue(getKey() + "/instanceCustomTimeId", instanceCustomTimeId);
    }

    public void setInstanceHour(@Nullable Integer instanceHour) {
        if (getInstanceHour() != null && getInstanceHour().equals(instanceHour))
            return;

        mInstanceJson.setInstanceHour(instanceHour);
        addValue(getKey() + "/instanceHour", instanceHour);
    }

    public void setInstanceMinute(@Nullable Integer instanceMinute) {
        if (getInstanceMinute() != null && getInstanceMinute().equals(instanceMinute))
            return;

        mInstanceJson.setInstanceMinute(instanceMinute);
        addValue(getKey() + "/instanceMinute", instanceMinute);
    }

    public void setDone(@Nullable Long done) {
        if (getDone() != null && getDone().equals(done))
            return;

        mInstanceJson.setDone(done);
        addValue(getKey() + "/done", done);
    }

    @Nullable
    public Double getOrdinal() {
        return mInstanceJson.getOrdinal();
    }

    public void setOrdinal(double ordinal) {
        if (getOrdinal() != null && getOrdinal().equals(ordinal))
            return;

        mInstanceJson.setOrdinal(ordinal);
        addValue(getKey() + "/ordinal", ordinal);
    }
}

File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteInstanceRecord.kt
package com.krystianwsul.checkme.firebase.records


import com.krystianwsul.checkme.utils.InstanceData
import com.krystianwsul.checkme.utils.ScheduleKey
import com.krystianwsul.common.ErrorLogger
import com.krystianwsul.common.domain.InstanceRecord
import com.krystianwsul.common.firebase.json.InstanceJson
import com.krystianwsul.common.time.Date
import com.krystianwsul.common.time.HourMinute
import com.krystianwsul.common.time.JsonTime
import com.krystianwsul.common.time.TimePair
import com.krystianwsul.common.utils.RemoteCustomTimeId
import kotlin.properties.Delegates.observable

class RemoteInstanceRecord<T : RemoteCustomTimeId>(
        create: Boolean,
        private val remoteTaskRecord: RemoteTaskRecord<T>,
        override val createObject: InstanceJson,
        val scheduleKey: ScheduleKey,
        private val firebaseKey: String,
        override val scheduleCustomTimeId: T?) : RemoteRecord(create), InstanceRecord<T> {

    companion object {

        private val hourMinuteRegex = Regex("^\\d\\d:\\d\\d$")

        private val hourMinuteKeyRegex = Regex("^(\\d\\d\\d\\d)-(\\d?\\d)-(\\d?\\d)-(\\d?\\d)-(\\d?\\d)$")
        private val customTimeKeyRegex = Regex("^(\\d\\d\\d\\d)-(\\d?\\d)-(\\d?\\d)-(.+)$")

        fun scheduleKeyToString(scheduleKey: ScheduleKey): String {
            var key = scheduleKey.scheduleDate.year.toString() + "-" + scheduleKey.scheduleDate.month + "-" + scheduleKey.scheduleDate.day + "-"
            key += scheduleKey.scheduleTimePair.let {
                if (it.customTimeKey != null) {
                    check(it.hourMinute == null)

                    it.customTimeKey!!.remoteCustomTimeId
                } else {
                    it.hourMinute!!.hour.toString() + "-" + it.hourMinute!!.minute
                }
            }

            return key
        }

        fun <T : RemoteCustomTimeId> stringToScheduleKey(remoteProjectRecord: RemoteProjectRecord<T>, key: String): Pair<ScheduleKey, T?> {
            val hourMinuteMatchResult = hourMinuteKeyRegex.find(key)

            fun MatchResult.getInt(position: Int) = groupValues[position].toInt()

            if (hourMinuteMatchResult != null) {
                val year = hourMinuteMatchResult.getInt(1)
                val month = hourMinuteMatchResult.getInt(2)
                val day = hourMinuteMatchResult.getInt(3)
                val hour = hourMinuteMatchResult.getInt(4)
                val minute = hourMinuteMatchResult.getInt(5)

                return Pair(ScheduleKey(Date(year, month, day), TimePair(HourMinute(hour, minute))), null)
            } else {
                val customTimeMatchResult = customTimeKeyRegex.find(key)
                checkNotNull(customTimeMatchResult)

                val year = customTimeMatchResult.getInt(1)
                val month = customTimeMatchResult.getInt(2)
                val day = customTimeMatchResult.getInt(3)

                val customTimeId = customTimeMatchResult.groupValues[4]
                check(customTimeId.isNotEmpty())

                val customTimeRecord = remoteProjectRecord.getCustomTimeRecord(customTimeId)

                val customTimeKey = remoteProjectRecord.getRemoteCustomTimeKey(customTimeRecord.projectId, customTimeId)

                return Pair(ScheduleKey(Date(year, month, day), TimePair(customTimeKey)), customTimeKey.remoteCustomTimeId)
            }
        }
    }

    override val key by lazy { remoteTaskRecord.key + "/instances/" + firebaseKey }

    val taskId by lazy { remoteTaskRecord.id }

    override var done: Long?
        get() = createObject.done
        set(done) {
            if (done == createObject.done)
                return

            createObject.done = done
            addValue("$key/done", done)
        }

    override val scheduleYear by lazy { scheduleKey.scheduleDate.year }
    override val scheduleMonth by lazy { scheduleKey.scheduleDate.month }
    override val scheduleDay by lazy { scheduleKey.scheduleDate.day }

    override val scheduleHour by lazy {
        scheduleKey.scheduleTimePair
                .hourMinute
                ?.hour
    }

    override val scheduleMinute by lazy {
        scheduleKey.scheduleTimePair
                .hourMinute
                ?.minute
    }

    private fun getInitialInstanceDate(): Date? {
        createObject.instanceDate
                .takeUnless { it.isNullOrEmpty() }
                ?.let { return Date.fromJson(it) }

        val instanceYear = createObject.instanceYear
        val instanceMonth = createObject.instanceMonth
        val instanceDay = createObject.instanceDay

        if (((instanceYear != null) != (instanceMonth != null)) || (instanceYear != null) != (instanceDay != null))
            ErrorLogger.instance.logException(InstanceData.InconsistentInstanceException("instance: " + remoteTaskRecord.key + " " + key + ", instanceYear: $instanceYear, instanceMonth: $instanceMonth, instanceDay: $instanceDay"))

        return if (instanceYear != null && instanceMonth != null && instanceDay != null)
            Date(instanceYear, instanceMonth, instanceDay)
        else
            null
    }

    override var instanceDate by observable(getInitialInstanceDate()) { _, _, value ->
        val instanceYear = value!!.year
        val instanceMonth = value.month
        val instanceDay = value.day

        if (instanceYear != createObject.instanceYear) {
            createObject.instanceYear = instanceYear
            addValue("$key/instanceYear", instanceYear)
        }

        if (instanceMonth != createObject.instanceMonth) {
            createObject.instanceMonth = instanceMonth
            addValue("$key/instanceMonth", instanceMonth)
        }

        if (instanceDay != createObject.instanceDay) {
            createObject.instanceDay = instanceDay
            addValue("$key/instanceDay", instanceDay)
        }

        val json = value.toJson()

        if (json != createObject.instanceDate) {
            createObject.instanceDate = json
            addValue("$key/instanceDate", json)
        }
    }

    private val initialInstanceJsonTime: JsonTime<T>?

    init {
        initialInstanceJsonTime = createObject.instanceTime
                ?.let {
                    val matchResult = hourMinuteRegex.find(it)
                    if (matchResult != null)
                        JsonTime.Normal<T>(HourMinute.fromJson(it))
                    else
                        JsonTime.Custom(remoteTaskRecord.getRemoteCustomTimeId(it))
                }
                ?: createObject.instanceCustomTimeId?.let { JsonTime.Custom(remoteTaskRecord.getRemoteCustomTimeId(it)) }
                        ?: createObject.instanceHour?.let { hour -> createObject.instanceMinute?.let { JsonTime.Normal<T>(HourMinute(hour, it)) } }
    }

    override var instanceJsonTime by observable(initialInstanceJsonTime) { _, _, value ->
        var customTimeId: T? = null
        var hourMinute: HourMinute? = null
        when (value) {
            is JsonTime.Custom -> customTimeId = value.id
            is JsonTime.Normal -> hourMinute = value.hourMinute
        }

        if (customTimeId != createObject.instanceCustomTimeId?.let { remoteTaskRecord.getRemoteCustomTimeId(it) }) {
            createObject.instanceCustomTimeId = customTimeId?.value
            addValue("$key/instanceCustomTimeId", customTimeId?.value)
        }

        if (hourMinute?.hour != createObject.instanceHour) {
            createObject.instanceHour = hourMinute?.hour
            addValue("$key/instanceHour", hourMinute?.hour)
        }

        if (hourMinute?.minute != createObject.instanceMinute) {
            createObject.instanceMinute = hourMinute?.minute
            addValue("$key/instanceMinute", hourMinute?.minute)
        }

        val instanceTime = value?.toJson()

        if (instanceTime != createObject.instanceTime) {
            createObject.instanceTime = instanceTime
            addValue("$key/instanceTime", instanceTime)
        }
    }

    override var ordinal
        get() = createObject.ordinal
        set(ordinal) {
            if (ordinal == createObject.ordinal)
                return

            createObject.ordinal = ordinal
            addValue("$key/ordinal", ordinal)
        }

    override var hidden
        get() = createObject.hidden
        set(value) {
            if (value == createObject.hidden)
                return

            createObject.hidden = value
            addValue("$key/hidden", value)
        }

    override fun deleteFromParent() = check(remoteTaskRecord.remoteInstanceRecords.remove(scheduleKey) == this)
}
